### ISSUE TITLE
Aleatoric v/s epistemic errors, + Error propagation

### ISSUE THREAD

**Issue by astrogilda** [stanfordmlgroup/ngboost#133]
Hey!

(1) Could you clarify if the 'sigma' that results from ngboost corresponds to the aleatoric, epistemic or their sum (i.e. sigma_ngboost = (sigma_epistemic**2 + sigma_aleatoric**2)**.5)? I have been reading about these two different types of errors, but am still not sure which is the one that's spit out by ngboost?

(2) If my independent variables (X) have associated uncertainties (say normal), because they are measurements from physical sensors, and I want to propagate these all the way through to the uncertainties on my output of interest (y), should I augment the dataset with samples drawn from the PDFs of X? Or is there another way to do this?

**Comment by alejandroschuler** [comment]
Hi @astrogilda, does #76 answer your question? Basically, NGBoost returns what it thinks is a model of how the world works. So if you observe a measurement of X=x, NGBoost is telling you the likelihood that Y takes a particular value y. That likelihood may or may not be accurate, because the model may or may not be correct. If you had refit the model using a slightly different dataset, the estimated likelihood you get back could be slightly different.

It's not clear to me what you're trying to accomplish in the second question. What is the quantity you are trying to estimate? I'm imagining something like: you have a "true" measurement Z of which X is a corrupted measurement. You somehow know what P(X|Z=z) is (normal with mean z and unit variance, for instance). You also have a dataset of outputs and corrupted measurements (y,x) that you can use. You would like to know the likelihood that the output Y is y if Z is z (in other words P(Y|Z=z)) perhaps because in the future you will have a more accurate sensor. Is that what you're trying to get at?

**Comment by astrogilda** [comment]
Hey @alejandroschuler! Thanks for such a quick response.

Yes, #76 answers my question :)

And yes, that is PRECISELY what I am trying to get at. Do you have any suggestions/solutions?

**Comment by alejandroschuler** [comment]
No problem @astrogilda. I don't have any bootstrapping heuristics... I'd keep increasing it until your uncertainty estimates with additional bootstrap samples are changing less than you care about.

About your second question- if that's what you're trying to do, I don't think it's possible. You simply don't have enough information. My argument is as follows:

![Screen Shot 2020-06-14 at 3 59 57 PM](https://user-images.githubusercontent.com/7213095/84606152-19b70000-ae58-11ea-9a8d-0faf58891264.png)

**Comment by astrogilda** [comment]
Thanks for the quick response, again!

I understand your argument above. But how does one square this with data-augmentation in deep learning, where we basically make multiple copies of a given observations/sample, by repeatedly sampling from the noise distribution that we want to marginalize over? Intuitively, why wouldn't that approach work here as well?

**Comment by alejandroschuler** [comment]
In the data augmentation setting, you have some "true" image/reality Z of which you receive a transformed version of (X) which is cropped, scaled, inverted, etc. It so happens that some of those transformations are invertible or composable, so you can take X and apply those transformations to arrive at other samples from P(X|Z) with the same label Y. The big difference between what you want to do and what's going on there is that you want P(Y|Z), whereas in data augmentation we're just using additional samples of X to better learn the relationship P(Y|X). There is ultimately no reference to the "true" image/reality Z that we imagine generates X because we never care about it and will never observe it.

**Comment by astrogilda** [comment]
Ah I see. So if I relax the requirement that I need P(Y|Z), by asserting
that we will never observe the true underlying distribution Z but only
corrupted samples of it (X), then can I augment my dataset the same way one
would do it in DL?

Best,
Sankalp Gilda

**Comment by alejandroschuler** [comment]
Yes, that's totally fine as long as you are sure that the two (original and augmented) "images" would still deterministically have the same "label".

**Comment by astrogilda** [comment]
Perfect, thanks!

A follow up question--in DL, one doesn't technically just add corrupted copies of a sample to the training dataset, but instead in each iteration samples from the underlying noisy distribution (e.g. flipped version of an image in epoch 0, rotated by 20 deg in epoch 1, rotated by 40 deg in epoch 2, and so on). What would the equivalent of that approach be here? Or is there none, and I am forced to simply increase the total size of the dataset many fold?

Finally, and this relates to #76: if I do end up increasing the size of the dataset, how should the bootstrapping be done in that case? Should I take special precautions to ensure in a given bag contains only one noisy observation of a sample?

**Comment by alejandroschuler** [comment]
There isn't functionality to do anything like that in ngboost, so you'd have to hack the codebase a little bit to get it to work like that. It is certainly possible. I'd introduce something like a "dataloader" that would generate a subsample of data at each boosting iteration. Would probably take a while to get used to the code and understand what you're doing so I'd guess not worth it for you unless you're looking for a project or desperately need that feature. 

As for bootstrapping.... hmm, I think I would take a bootstrap sample from your dataset before doing anything else at all, then feed that sample through your whole pipeline (data augmentation, learning, etc.), then repeat the process for another bootstrap sample, etc. The idea is to perfectly replicate the entirety of what you're doing, but start with different data that you could have sampled if you were doing a perfect replication of the experiment.

**Comment by astrogilda** [comment]
That certainly looks like a non-trivial task. While I'd love to do things right, unfortunately I can't really get into it at the moment, on account of being tied up with a few other projects. Is there a hacky way to get 'close enough' answers? For instance, if I don't create a dataloader and load from it at every boosting iteration, but instead just add corrupted copies of an observation (since I know the underlying noise generation function) to the total training set (so a training set of 1000 becomes 5000 if I add 5 corrupted copies of each sample), would that do the trick? What kinds of biases would I be subject to in that case?

'The idea is to perfectly replicate the entirety of what you're doing, but start with different data that you could have sampled if you were doing a perfect replication of the experiment.'--Could you please explain this in other words? I didn't quite follow.

**Comment by alejandroschuler** [comment]
Yes, just adding examples to the existing dataset would work fine, at the cost of having more data to handle and thus slower compute. But the computer's time is worth less than yours, so probably a good trade-off :)

Sure, let me try again. Let's say we're estimating the expected value of Y when X=4. To do that, I'm going to take my dataset, which consists of n samples of (x,y) and push it through some algorithm (ngboost) that will give me a model that estimates P(Y|X=x). Before doing that, we can also do whatever data augmentation we like. Now, since we have the conditional density P(Y|X=x), we can take the expectation of it to get y = E[Y|X=x]. Great. Now we have an expected value (y) and also a prediction interval around y given by finding where 95% of the mass of P(Y|X=x) falls on the domain of y. That is a measure of uncertainty as provided by the model. 

However, as we have discussed, the model isn't going to be right every time. The model gives back an **estimate** of P(Y|X). There is uncertainty in the model's estimate of the mean and in the model's estimate of the uncertainty because the data we started with were some random sample from the distribution P(Y,X). If we were to repeat our analysis by taking another dataset of the same size and fitting P(Y|X=x), we would get a slightly different expected value y' at X=4 and a slightly different prediction interval. The estimates we have are therefore themselves random variables.

One way to estimate how much variability there is in our estimates (of the mean and prediction interval) is to repeat the analysis with another dataset of the same size multiple times. But we don't have infinite data to do this. So what we do is pretend that P(Y,X) is given by the "empirical distribution" of the data we have (i.e. sampling from replacement from [(x1, y1), ... (xn, yn)]). Call the empirical distribution P*(X,Y). So instead of sampling another dataset from the real world P(X,Y) and repeating our analysis, we're going to sample a dataset from our empirical distribution P*(X,Y) and pretend that's a sample from P(X,Y). We re-do our analysis on this sample and get a different estimate for y, and we do this many (B) times so we end up with a B estimates of the expectation at X=4, call them [y*1, y*2, ... y*B] and B prediction intervals. 

That lets us directly investigate the variability in our estimates. Perhaps in our real data, y=1 with prediction interval from 0.5 to 1.5. But across the bootstrap replicates, y often varied from 0.9 to 1.1 and the prediction interval often spanned from 0 to 2, etc. The bootstrap replicates of the estimates are essentially a stand-in for the sampling distributions of these quantities.

The point is that the source of randomness is your initial data sampling process that led to you having the dataset that you have in your hands. So if you want to measure that randomness, you need to simulate the process that led to you having those data. Everything that comes after you collected the data is just part of your analysis, even if it's nondeterministic.

**Comment by astrogilda** [comment]
Thanks for the indepth answer!

So please correct me if I am wrong, but haven't you simply described the process of bootstrapping? I am specifically curious about HOW to create the bags in bagging, given the fact that the input samples themselves have uncertainties. Though I wonder now, if this is even the correct way to approach this. Allow me to explain a bit more.

Aleatoric uncertainty is the one that arises due to stochasticity in the dataset, and an absence of all explanatory variables. For example, if one is trying to predict the digit in CIFAR10, and instead of providing the model access to complete images, we crop out the right half of all of them, then the uncertainty in the model's predictions would be aleatoric in nature, since they won't reduce by adding more samples (n), but will do when provided access to the right half of the images (i.e. more features, d). I believe that the uncertainty that I am trying to capture on account of noisy input samples is also aleatoric, since it arises due to stochasticity in inputs. Would you agree with this?

If so, then the estimation of this uncertainty should be independent of the bagging/boostrapping process, since that is only required to assess epistemic uncertainty, as you explain in your blog post linked above. In which case, I would have to ammend my question, and instead ask--what is the best way to do cross validation (and not bagging) to account for this uncertainty. Below, the table at the top is the original input dataset, where say we decide to use the first two rows in training and the last one in validation. The table below it is contains a corrupted version of the observations (but with the same output), again with the first two rows to be used for training and the last one for validation. Now, if one trains on the first dataset, and obtains some mu and sigma for the third row (the validation set) using ngboost, and repeats the exercise to obtain mu and sigma for the third row from the second dataset, what would be a statistically 'sound' way to combine these results?

(Sorry for the wall of text--the very act of writing this out is making me rething the way I was approaching the problem!) 

Feature1 Feature2 Feature3 Output
10                20            30          4
12                22            32          5
14                24            34          6

**Comment by alejandroschuler** [comment]
I don't know if "aleatoric" vs "epistemic" uncertainty is the right way to look at this. I'm thinking concretely: as we've shown, data augmentation is just a thing we do to get a (presumably) better estimate of P(Y|X). As such, it's just part of a mapping from the data to the estimate of interest, and thus should be included within the bootstrapping loop with no special consideration.
